ParseException message should contain the path of the broken .editorconfig file

Whenever `.editorconfig` parsing exception happens, library throws `ParseException`:
```java
Caused by: org.ec4j.core.parser.ParseException: Equals sign '==' missing after property name '' at 1:1 (0)
        at org.ec4j.core.parser.ErrorHandler$3.error(ErrorHandler.java:54)
        at org.ec4j.core.parser.EditorConfigParser.readProperty(EditorConfigParser.java:324)
        at org.ec4j.core.parser.EditorConfigParser.readLine(EditorConfigParser.java:180)
        at org.ec4j.core.parser.EditorConfigParser.readLines(EditorConfigParser.java:148)
        at org.ec4j.core.parser.EditorConfigParser.parse(EditorConfigParser.java:131)
```

Currently there is no way to tell which exactly `.editorconfig` file failed.

Could you add to `ParseException` path to failed file?

Originally reported as a bug in ktlint project: https://github.com/pinterest/ktlint/issues/790 and we could not print more user-friendly message in this case.

--- a/core/src/main/java/org/ec4j/core/parser/AbstractValidatingHandler.java
+++ b/core/src/main/java/org/ec4j/core/parser/AbstractValidatingHandler.java
@@ -48,9 +48,11 @@ public abstract class AbstractValidatingHandler implements EditorConfigHandler {
         final Glob glob = new Glob(globSource);
         final PatternSyntaxException e = glob.getError();
         if (e != null) {
-            final String msg = String.format("The glob '%s' is not valid: %s", globSource, e.getMessage());
+            final String msg = String.format("The glob '%s' is not valid: %s",
+                    globSource,
+                    e.getMessage());
             context.getErrorHandler().error(context,
-                    new ErrorEvent(globStart, context.getLocation(), msg, ErrorType.INVALID_GLOB));
+                    new ErrorEvent(globStart, context.getLocation(), context.getResource(), msg, ErrorType.INVALID_GLOB));
         }
         glob(context, glob);
         globStart = null;
@@ -74,8 +76,13 @@ public abstract class AbstractValidatingHandler implements EditorConfigHandler {
     public void endPropertyValue(ParseContext context, String value) {
         final PropertyValue<?> propValue = type == null ? PropertyValue.valid(value, value) : type.parse(value);
         if (!propValue.isValid()) {
-            context.getErrorHandler().error(context, new ErrorEvent(propertyValueStart, context.getLocation(),
-                    propValue.getErrorMessage(), ErrorType.INVALID_PROPERTY_VALUE));
+            context.getErrorHandler().error(context,
+                    new ErrorEvent(
+                            propertyValueStart,
+                            context.getLocation(),
+                            context.getResource(),
+                            propValue.getErrorMessage(),
+                            ErrorType.INVALID_PROPERTY_VALUE));
         }
         propertyValue(context, propValue);
         this.type = null;

--- a/core/src/main/java/org/ec4j/core/parser/EditorConfigParser.java
+++ b/core/src/main/java/org/ec4j/core/parser/EditorConfigParser.java
@@ -130,8 +130,12 @@ public class EditorConfigParser implements ParseContext {
             this.reader = reader;
             readLines();
             if (!isEndOfText()) {
-                Location location = getLocation();
-                ErrorEvent e = new ErrorEvent(location, location, "Found unexpected character; expected end of input",
+                final Location location = getLocation();
+                final ErrorEvent e = new ErrorEvent(
+                        location,
+                        location,
+                        resource,
+                        "Found unexpected character; expected end of input",
                         ErrorType.EXPECTED_END_OF_INPUT);
                 errorHandler.error(this, e);
             }
@@ -210,7 +214,11 @@ public class EditorConfigParser implements ParseContext {
 
     private void globNotClosed() {
         final Location location = getLocation();
-        ErrorEvent e = new ErrorEvent(location, location, "Glob pattern not closed. Expected ']'",
+        final ErrorEvent e = new ErrorEvent(
+                location,
+                location,
+                resource,
+                "Glob pattern not closed. Expected ']'",
                 ErrorType.GLOB_NOT_CLOSED);
         errorHandler.error(this, e);
     }
@@ -264,12 +272,20 @@ public class EditorConfigParser implements ParseContext {
              */
             if (isEndOfText()) {
                 final Location location = getLocation();
-                ErrorEvent e = new ErrorEvent(location, location, "Unexpected end of input",
+                ErrorEvent e = new ErrorEvent(
+                        location,
+                        location,
+                        resource,
+                        "Unexpected end of input",
                         ErrorType.UNEXPECTED_END_OF_INPUT);
                 errorHandler.error(this, e);
             } else if (current < 0x20) {
                 final Location location = getLocation();
-                ErrorEvent e = new ErrorEvent(location, location, "Expected a valid string character",
+                ErrorEvent e = new ErrorEvent(
+                        location,
+                        location,
+                        resource,
+                        "Expected a valid string character",
                         ErrorType.EXPECTED_STRING_CHARACTER);
                 errorHandler.error(this, e);
             } else {
@@ -320,8 +336,11 @@ public class EditorConfigParser implements ParseContext {
         skipWhiteSpace();
         if (!readChar('=') && !readChar(':')) {
             final Location location = getLocation();
-            ErrorEvent e = new ErrorEvent(location, location,
-                    "Equals sign '==' missing after property name '" + name + "'",
+            ErrorEvent e = new ErrorEvent(
+                    location,
+                    location,
+                    resource,
+                    "Equals sign '=' missing after property name '" + name + "'",
                     ErrorType.PROPERTY_ASSIGNMENT_MISSING);
             errorHandler.error(this, e);
         }
@@ -331,7 +350,11 @@ public class EditorConfigParser implements ParseContext {
         String value = readString(StopReading.PropertyValue);
         if (value.length() < 1) {
             final Location location = getLocation();
-            ErrorEvent e = new ErrorEvent(location, location, "Property '" + name + "' has no value",
+            ErrorEvent e = new ErrorEvent(
+                    location,
+                    location,
+                    resource,
+                    "Property '" + name + "' has no value",
                     ErrorType.PROPERTY_VALUE_MISSING);
             errorHandler.error(this, e);
         }

--- a/core/src/main/java/org/ec4j/core/parser/ErrorEvent.java
+++ b/core/src/main/java/org/ec4j/core/parser/ErrorEvent.java
@@ -16,6 +16,7 @@
  */
 package org.ec4j.core.parser;
 
+import org.ec4j.core.Resource;
 import org.ec4j.core.model.Glob;
 import org.ec4j.core.model.PropertyType;
 
@@ -64,8 +65,11 @@ public class ErrorEvent extends Span {
 
     private final String message;
 
-    public ErrorEvent(Location start, Location end, String message, ErrorType errorType) {
+    private final Resource resource;
+
+    public ErrorEvent(Location start, Location end, Resource resource, String message, ErrorType errorType) {
         super(start, end);
+        this.resource = resource;
         this.message = message;
         this.errorType = errorType;
     }
@@ -84,6 +88,13 @@ public class ErrorEvent extends Span {
      */
     public String getMessage() {
         return message;
+    }
+
+    /**
+     * @return the {@link Resource} where this {@link ErrorEvent} occurred
+     */
+    public Resource getResource() {
+        return resource;
     }
 
 }

--- a/core/src/main/java/org/ec4j/core/parser/ParseException.java
+++ b/core/src/main/java/org/ec4j/core/parser/ParseException.java
@@ -28,7 +28,12 @@ public class ParseException extends RuntimeException {
     private static final long serialVersionUID = -3857553902141819279L;
 
     public ParseException(ErrorEvent errorEvent) {
-        super(errorEvent.getMessage() + " at " + errorEvent.getStart());
+        super(String.format(
+                "%s:%d:%d: %s",
+                errorEvent.getResource(),
+                errorEvent.getStart().getLine(),
+                errorEvent.getStart().getColumn(),
+                errorEvent.getMessage()));
     }
 
 }

--- a/core/src/test/java/org/ec4j/core/EditorConfigParserTest.java
+++ b/core/src/test/java/org/ec4j/core/EditorConfigParserTest.java
@@ -23,6 +23,7 @@ import java.util.Iterator;
 
 import org.ec4j.core.Resource.Resources.StringResourceTree;
 import org.ec4j.core.model.Property;
+import org.ec4j.core.parser.ParseException;
 import org.junit.Assert;
 import org.junit.Test;
 
@@ -256,4 +257,22 @@ public class EditorConfigParserTest {
                 .queryProperties(tree.getResource(testFile)).getProperties().values();
         Assert.assertEquals(0, properties.size());
     }
+
+    @Test
+    public void unfinishedGlob() throws IOException {
+        final String testFile = "root/test4";
+        StringResourceTree tree = StringResourceTree.builder() //
+                .resource("root/.editorconfig", "[foo")//
+                .touch(testFile) //
+                .build();
+
+        try {
+            ResourcePropertiesService.default_()
+                    .queryProperties(tree.getResource(testFile)).getProperties().values();
+            Assert.fail("ParseException expected");
+        } catch (ParseException e) {
+            Assert.assertEquals(e.getMessage(), "string:root/.editorconfig:1:3: Glob pattern not closed. Expected ']'");
+        }
+    }
+
 }

